wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto3WireProtocCompatibilityTests.kt

```diff
@@ -40,4 +52,221 @@ class Proto3WireProtocCompatibilityTests {
+        .add(PizzaOuterClass.BuyOneGetOnePromotion.getDescriptor())
+        .add(PizzaOuterClass.FreeGarlicBreadPromotion.getDescriptor())
+        .build()
+
+    val jsonPrinter = JsonFormat.printer()
+        .usingTypeRegistry(typeRegistry)
+    val protocMessageJson = jsonPrinter.print(protocMessage)
+
+    val wireMessage = PizzaDelivery(
+        address = "507 Cross Street",
+        pizzas = listOf(Pizza(toppings = listOf("pineapple", "onion"))),
+        promotion = AnyMessage.pack(BuyOneGetOnePromotion(coupon = "MAUI"))
+    )
+
+    val moshi = Moshi.Builder()
+        .add(WireJsonAdapterFactory()
+            .plus(BuyOneGetOnePromotion.ADAPTER, FreeGarlicBreadPromotion.ADAPTER))
+        .build()
+    val jsonAdapter = moshi.adapter(PizzaDelivery::class.java)
+    val moshiMessageJson = jsonAdapter.toJson(wireMessage)
+
+    // Parsing the two json because this should be order insensitive.
+    assertJsonEquals(moshiMessageJson, protocMessageJson)
+
+    // Now each parses the other's json.
+    val jsonParser = JsonFormat.parser().usingTypeRegistry(typeRegistry)
+    val protocMessageDecodedFromWireJson = PizzaOuterClass.PizzaDelivery.newBuilder()
+        .apply { jsonParser.merge(moshiMessageJson, this) }
+        .build()
+
+    val wireMessageDecodedFromProtocJson = jsonAdapter.fromJson(protocMessageJson)
+
+    assertThat(protocMessageDecodedFromWireJson).isEqualTo(protocMessage)
+    assertThat(wireMessageDecodedFromProtocJson).isEqualTo(wireMessage)
+  }
+
+  @Test fun unregisteredTypeOnReading() {
+    val moshi = Moshi.Builder()
+        .add(WireJsonAdapterFactory())
+        .build()
+
+    val jsonAdapter = moshi.adapter(PizzaDelivery::class.java)
+
+    val json = """
+        |{
+        |  "address": "507 Cross Street",
+        |  "pizzas": [
+        |    {
+        |      "toppings": [
+        |        "pineapple",
+        |        "onion"
+        |      ]
+        |    }
+        |  ],
+        |  "promotion": {
+        |    "@type": "type.googleapis.com/squareup.proto3.pizza.BuyOneGetOnePromotion",
+        |    "coupon": "MAUI"
+        |  }
+        |}
+        """.trimMargin()
+
+    try {
+      jsonAdapter.fromJson(json)
+      fail()
+    } catch (expected: IOException) {
+      assertThat(expected).hasMessage("Cannot resolve type: " +
+          "type.googleapis.com/squareup.proto3.pizza.BuyOneGetOnePromotion")
+    }
+  }
+
+  @Test fun unrecognizedTypeOnWriting() {
+    val pizzaDelivery = PizzaDelivery(
+        address = "507 Cross Street",
+        pizzas = listOf(Pizza(toppings = listOf("pineapple", "onion"))),
+        promotion = AnyMessage.pack(BuyOneGetOnePromotion(coupon = "MAUI"))
+    )
+
+    val moshi = Moshi.Builder()
+        .add(WireJsonAdapterFactory())
+        .build()
+
+    val jsonAdapter = moshi.adapter(PizzaDelivery::class.java)
+    try {
+      jsonAdapter.toJson(pizzaDelivery)
+      fail()
+    } catch (expected: AssertionError) {
```

Comment:
What? This is definitely the wrong type to be throwing

Reply:
`fail` also throws this so you never would actually fail!

Reply:
hmmhmmhhm
https://github.com/square/moshi/blob/master/moshi/src/main/java/com/squareup/moshi/JsonAdapter.java#L62

Reply:
Right but `JsonDataException` is runtime so it doesn't get wrapped anymore. Cool.